### ISSUE TITLE
RFC: Expand Booking E2E tests

### ISSUE THREAD

**Issue by zomars** [calcom/cal.com#2255]
Since there's a ton of possible event type configurations we should expand our testing with some of the most common use cases, these are the ones that come to my mind:

- Can book an event with Zoom Video as location
- Can book an event with Google Meet as location
- Can book an event with Daily.co Video as location
- Can book an event with Jitsi Meet as location
- Can book an event with Huddle01 Video as location
- Can book an event with Link Meeting as location
- Can book an event with In Person as location
- Can book an event with Phone as location
- Can book an event with Additional Fields
- Can book an event with Web3 enabled (This is a tricky one with playwright)

Expanding testing for these will allow us to make changes with peace of mind that none of the critical paths are breaking. I can tell from experience with App Store. 

Knowing this, there are more questions on how to achieve this.

- How can we save/reuse/mock testing credentials for each of the needed integrations?
- Can we mock it enough to have confidence in the booking not breaking?

Will expand as soon as I have more information....

**Comment by zomars** [comment]
@hariombalhara @denik1981 I would love to hear your input on this 🙏🏽

**Comment by PeerRich** [comment]
great RFC

**Comment by hariombalhara** [comment]
I have added zoom tests in PR #2202 

It right now just tests that user is able to install zoom app. We can very easily expand its to Event creation. 

The philosophy I am using here is that third party code shouldn't be tested in core. We should just mock the API as per their documentation and use those mocks.(Core Tests that run on every PR) 
Benefits:
- You can create any type of third party API scenario. With actual credentials it would almost be impossible. 
- You can test with any version of API(Alpha, Beta, Deprecated, Stable) 
- Tests would be faster to execute. 
- Flaky Test won't be there. 
- You understand the documentation well and are able to quickly add cose for all scenarios. 

But we must test atleast one happy path with each third party integration. (Extended Tests that run once or at scheduled time every day on main)

**Comment by zomars** [comment]
Yeah, we already use Checkly to run playwright on staging/production. We could setup those tests in there and use the example account with all integrations enabled.

**Comment by denik1981** [comment]
Thanks for taking me into consideration. 
Sure, I will take a look and I will be happy to contribute to this.

> Expanding testing for these will allow us to make changes with peace of mind that none of the critical paths are breaking. I can tell from experience with App Store.

Indeed and this statement can be expanded to other areas of Cal. Especially, when it counts to moving things around like the major PRs with turbo and now with the app-store, or even with a small chore code modification is when you want to have a solid base of tests holding your back and help you be certain you didn't break anything else. 

@hariombalhara I love **MSW**. It is pure gold that library. The only concern I have with MSW is that it lacks a very important feature that I can find, for eg,  in NOCK which is the ability to mock my response based on a real response from the third party we are mocking. Other than that, I pretty much love MSW. I agree with you that mocking will allow us to test without having reliability issues in the third-party resource and to reduce the latency of tripping back and forth when using a real network connection if we can resolve everything locally. Also third-party resources most likely will have a threshold that will prevent us from flooding their endpoints which high request rates which is a common thing in the testing stage. I mean, of course we can debounce the request, but in the end this will scale with the added cost of having slow e2e tests. I would take into consideration using MSW.

**Comment by denik1981** [comment]
Quick update: I'm working on this and I think I find a good solution to do both. Test again real and mocked data in demand although I need to explore this further and, of course, test it. Will see.

@zomars Could you please review #2293? 
For my app-store tests I will be using a users fixtures I have created in this PR, thus I'm rebasing from it.

**Comment by miguelnietoa** [comment]
Hello everybody, I would like to collaborate on some E2E bookings tests. 
I think you can create an Epic that lists all the possible use cases and work on them, just a comment.
